modules/mysql/main.tf

```diff
@@ -0,0 +1,194 @@
+ * limitations under the License.
+ */
+
+locals {
+  default_tier                            = "db-n1-standard-1"
+  default_activation_policy               = "ALWAYS"           // ON_DEMAND
+  default_crash_safe_replication          = true
+  default_disk_autoresize                 = true
+  default_disk_size                       = 10
+  default_disk_type                       = "PD_SSD"
+  default_pricing_plan                    = "PER_USE"
+  default_maintenance_window_day          = 1
+  default_maintenance_window_hour         = 23
+  default_replication_type                = "SYNCHRONOUS"
+  default_maintenance_window_update_track = "canary"
+
+  default_user_host = "%"
+
+  primary_zone       = "${lookup(var.master, "zone")}"
+  read_replica_zones = ["${compact(split(",", lookup(var.read_replica, "zones", "")))}"]
+
+  zone_mapping = {
+    enabled  = ["${local.read_replica_zones}"]
+    disabled = "${list(local.primary_zone)}"
+  }
+
+  zones_enabled = "${length(local.read_replica_zones) > 0}"
+  mod_by        = "${local.zones_enabled ? length(local.read_replica_zones) : 1}"
+
+  zones = "${local.zone_mapping["${local.zones_enabled ? "enabled" : "disabled"}"]}"
+}
+
+resource "google_sql_database_instance" "master" {
+  project          = "${var.project_id}"
+  name             = "${var.name}"
+  database_version = "${var.database_version}"
+  region           = "${var.region}"
+
+  settings {
+    tier                        = "${lookup(var.master, "tier", local.default_tier)}"
+    activation_policy           = "${lookup(var.master, "activation_policy", local.default_activation_policy)}"
+    authorized_gae_applications = ["${var.authorized_gae_applications}"]
+    backup_configuration        = ["${var.backup_configuration}"]
+    ip_configuration            = ["${var.ip_configuration}"]
+
+    disk_autoresize = "${lookup(var.master, "disk_autoresize", local.default_disk_autoresize)}"
+
+    disk_size      = "${lookup(var.master, "disk_size", local.default_disk_size)}"
+    disk_type      = "${lookup(var.master, "disk_type", local.default_disk_type)}"
+    pricing_plan   = "${lookup(var.master, "pricing_plan", local.default_pricing_plan)}"
+    user_labels    = "${var.master_labels}"
+    database_flags = ["${var.master_database_flags}"]
+
+    location_preference {
+      zone = "${var.region}-${lookup(var.master, "zone")}"
+    }
+
+    maintenance_window {
+      day          = "${lookup(var.master, "maintenance_window_day", local.default_maintenance_window_day)}"
+      hour         = "${lookup(var.master, "maintenance_window_hour", local.default_maintenance_window_hour)}"
+      update_track = "${lookup(var.master, "maintenance_window_update_track", local.default_maintenance_window_update_track)}"
+    }
+  }
+
+  lifecycle {
+    ignore_changes = ["disk_size"]
+  }
+}
+
+resource "google_sql_database_instance" "replicas" {
+  count                 = "${lookup(var.read_replica, "length", 0)}"
```

Comment:
It looks like read_replica is a map, not an array? So I think this logic is faulty - instead we should just have a variable which specifies whether a read replica is desired or not (HA)?

Reply:
Nope, this works fine. As you said, `read_replica` is map. But if read replica is not required, length will be 0 by default because the developer won't specify the `read_replica` argument.
`read_replica`'s arguments will be used for all read replicas as settings, only `zones` argument will be used for rotation.

Receiving your comment, I just added a few arguments for improving readability, they're `failover_replica`, `read_replica_size`.
The `failover_replica` is expected to set boolean. The `read_replica_size` is same with the `length` argument inside the `read_replicas` block.

Please review again.

Reply:
@morgante Is this fine?